providers/efa/verbs.c

```diff
@@ -1158,19 +1161,34 @@ static void efa_setup_qp(struct efa_context *ctx,
 			 struct ibv_qp_cap *cap,
 			 size_t page_size)
 {
-	uint16_t rq_desc_cnt;
+	uint32_t rq_desc_cnt_u32;
+	uint64_t sq_desc_cnt;
+	uint64_t rq_desc_cnt;
 
 	efa_qp_init_indices(qp);
 
-	qp->sq.wq.wqe_cnt = roundup_pow_of_two(max_t(uint32_t, cap->max_send_wr,
-						     ctx->min_sq_wr));
+	sq_desc_cnt = roundup_pow_of_two(max_t(uint32_t, cap->max_send_wr, ctx->min_sq_wr));
+	if (sq_desc_cnt > QP_DESC_COUNT_MAX) {
```

Comment:
How is it possible?

You call to efa_setup_qp() after successful efa_check_qp_limits().
In the latter function you have something like that:

  1217         if (attr->cap.max_send_wr > ctx->max_sq_wr) {
  1218                 verbs_err(&ctx->ibvctx,
  1219                           "Max send WR %u > %u\n", attr->cap.max_send_wr,
  1220                           ctx->max_sq_wr);
  1221                 return EINVAL;
  1222         }
  1223 
  1224         if (attr->cap.max_recv_wr > ctx->max_rq_wr) {
  1225                 verbs_err(&ctx->ibvctx,
  1226                           "Max receive WR %u > %u\n", attr->cap.max_recv_wr,
  1227                           ctx->max_rq_wr);
  1228                 return EINVAL;
  1229         }

I expect that your FW provides proper max_sq_wr and min_sq_wr values from the beginning.
If it is due to FW limitation, it needs to be fixed in the kernel. If there is no such limitation, this patch will be redundant,

Reply:
@rleon ,
Normally, as you say, it wouldn't be a problem ==> This is a fix for an issue found in our provider/efa code when bypassing the efa_check_qp_limits() with invalid values.

Reply:
How did you bypass efa_check_qp_limits()?
Do you have such flow in the driver?
Why don't you "fix" your efa_check_qp_limits?

Reply:
Bypassed efa_check_qp_limits() as part of a test.

Simply:
static int efa_check_qp_limits(struct efa_context *ctx,
			       struct ibv_qp_init_attr_ex *attr)
{
	**if (bypass_enable)
		return 0;**

	if (attr->cap.max_send_sge > ctx->max_sq_sge) {
		verbs_err(&ctx->ibvctx,
			  "Max send SGE %u > %u\n", attr->cap.max_send_sge,
			  ctx->max_sq_sge);
		return EINVAL;
	}

Reply:
You "deleted" very important part of QP initialization flow and now expect that other providers/efa functions will work correctly. I don't understand, why do you expect from us to merge it?

Reply:
@rleon , This code is a safer initialization sequence of the sq and rx descriptors count. Why not merging it?

Reply:
> @rleon , This code is a safer initialization sequence of the sq and rq descriptors count. Why not merging it?

Because this function doesn't exist as standalone and part of proper QP initialization flow. In that flow, QP validation is an important part that you can't skip.

Writing code for something that can't be is bad as it hurts readability later.

Reply:
roundup_pow_of_two() returns a uint64_t:
static inline uint64_t roundup_pow_of_two(uint64_t n)

The 64-bit result is assigned below to a 16-bit variable in the original code:
	uint16_t rq_desc_cnt;

rq_desc_cnt = roundup_pow_of_two(cap->max_recv_sge * cap->max_recv_wr);

This code will yield a compilation warning if we enable a higher level of compilation warning.

Note also that (cap->max_recv_sge * cap->max_recv_wr) could be larger than MAX_UINT16 (== 0xFFFF) in case cap->max_recv_sge and cap->max_recv_wr are updated ==> This is unrelated to the test that revealed the issue.

How is this code less readable when it has more comments than the previous code? :)

Reply:
> roundup_pow_of_two() returns a uint64_t: static inline uint64_t roundup_pow_of_two(uint64_t n)
> 
> The 64-bit result is assigned below to a 16-bit variable in the original code: uint16_t rq_desc_cnt;
> 
> rq_desc_cnt = roundup_pow_of_two(cap->max_recv_sge * cap->max_recv_wr);
> 
> and in this line, to a 32-bit variable (qp->sq.wq.wqe_cnt): qp->sq.wq.wqe_cnt = roundup_pow_of_two(max_t(uint32_t, cap->max_send_wr, ctx->min_sq_wr));

Everything above should be fixed by returning proper value from FW and having right qp_vaildate_limit function.

> 
> Both lines will yield a compilation warning if we enable a higher level of compiler warning.

But we don't enable.

> 
> Note also that (cap->max_recv_sge * cap->max_recv_wr) could be larger than MAX_UINT16 (== 0xFFFF) in case cap->max_recv_sge and cap->max_recv_wr are updated ==> This is unrelated to the test that revealed the issue.
> 
> How is this code less readable when it has more comments than the previous code? :)

It gives false assumption that such failure is real and needs to be treated specially.

Reply:
So how about adding another function in utils.h and use it in provider/efa/verbs.c:

#define MAX_LEADING_ZEROS (sizeof(uint32_t) * 8)
static inline uint32_t roundup_pow_of_two_32bit(uint32_t n)
{
        uint32_t shifts = (MAX_LEADING_ZEROS  - __builtin_clz(n));
        if (shifts == MAX_LEADING_ZEROS)
               shifts = MAX_LEADING_ZEROS - 1;
        return 1UL << shifts;
}

Reply:
OK, so how about the following code (to handle only the uint32_t issue):
static void efa_setup_qp(struct efa_context *ctx,
			 struct efa_qp *qp,
			 struct ibv_qp_cap *cap,
			 size_t page_size)
{
	uint32_t rq_desc_cnt_u32;

	efa_qp_init_indices(qp);

	**qp->sq.wq.wqe_cnt = (uint32_t)roundup_pow_of_two(max_t(uint32_t, cap->max_send_wr, ctx->min_sq_wr));**
	qp->sq.wq.max_sge = cap->max_send_sge;
	qp->sq.wq.desc_mask = qp->sq.wq.wqe_cnt - 1;

	qp->rq.wq.max_sge = cap->max_recv_sge;
	**rq_desc_cnt_u32 = (uint32_t)roundup_pow_of_two(cap->max_recv_sge * cap->max_recv_wr);**
	qp->rq.wq.desc_mask = rq_desc_cnt_u32 - 1;
	qp->rq.wq.wqe_cnt = rq_desc_cnt_u32 / qp->rq.wq.max_sge;

	qp->page_size = page_size;
}

Reply:
> 

1. We can add a check for 0.
2. The function returns uint32_t.

Reply:
> I don't think so.
> 
>     1. The result of __builtin_clz is undefined if n == 0.
> 
>     2. You are returning unsigned long, when uint32_t is unsigned int.

Also, regarding the previous proposed scheme,
1.  We can add a check for 0.
2. The function returns a uint32_t.

Reply:
> The question is, if this scheme looks like something you can accept...?

I'm missing an answer on "why should I merge it?".

Thanks

Reply:
On the receive queue: 
qp->rq.wq.desc_mask defined as a uint16.
uint16_t efa_wq::desc_mask

On our NIC, we have:
ctx->max_rq_wr 32768
ctx->max_rq_sge 3

==> The following condition will pass with no error if attr->cap.max_recv_wr == 32767 and max_recv_sge == 3
32767 * 3 = 98301 > MAX(uint16) ==> **It will yield an overflow error**.
```
	if (attr->cap.max_recv_sge > ctx->max_rq_sge) {
		verbs_err(&ctx->ibvctx,
			  "Max receive SGE %u > %u\n", attr->cap.max_recv_sge,
			  ctx->max_rq_sge);
		return EINVAL;
	}

	if (attr->cap.max_recv_wr > ctx->max_rq_wr) {
		verbs_err(&ctx->ibvctx,
			  "Max receive WR %u > %u\n", attr->cap.max_recv_wr,
			  ctx->max_rq_wr);
		return EINVAL;
	}
```

On the send queue, we have:
On the send queue:
We have uint32_t efa_wq::wqe_cnt and 
ctx->max_sq_wr 4096

Indeed, in this scenario there wouldn't be an issue for most values.